cli/src/commands/mod.rs

```diff
@@ -2376,6 +2454,127 @@ Please use `jj new 'all:x|y'` instead of `jj new --allow-large-revsets x y`.",
     Ok(())
 }
 
+fn cmd_next(ui: &mut Ui, command: &CommandHelper, args: &NextArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let edit = args.edit;
+    let amount = args.amount;
+    let current_wc_id = workspace_command
+        .get_wc_commit_id()
+        .ok_or_else(|| user_error("This command requires a working copy"))?;
+    let current_wc = workspace_command.repo().store().get_commit(current_wc_id)?;
+    let current_short = short_commit_hash(current_wc.id());
+    // If we're editing, start at the working-copy commit.
+    // Everything else starts from our direct parent.
+    let start_id = if edit {
+        current_wc_id
+    } else {
+        match current_wc.parent_ids() {
+            [parent_id] => parent_id,
+            _ => return Err(user_error("Cannot run `jj next` on a merge commit")),
+        }
+    };
+    let target_expression = RevsetExpression::commit(start_id.clone())
+        .descendants_at(amount)
+        // Negate the current working copy.
+        .minus(&RevsetExpression::commit(current_wc_id.clone()));
```

Comment:
I think if you have a linear chain of commits A through F and you're currently editing C, then `jj next` would error out because there's no child of B (`start_id`) that's C (`current_wc_id`). I think that's good. However, if the user instead did `jj next 2`, it would not error out because we don't subtract `D` from the set. So, I think what we want is instead this:

```
let target_revset = if edit {
  RevsetExpression::commit(start_id.clone()).descendants_at(amount)
} else {
  RevsetExpression::commit(start_id.clone()).descendants_at(amount)
        .minus(&RevsetExpression::commit(current_wc_id.clone().descendants()));
};
```

Feel free to extract `RevsetExpression::commit(start_id.clone()).descendants_at(amount)` to a variable if you can think of a good name :)

Reply:
extracted `find_descendendant_at` and `find_ancestor_at`.

Reply:
This doesn't seem done (the `.descendants()` part)

Reply:
Ditto for `prev`, so subtract the `ascendants()` ?

Reply:
I don't think we'll need it there because the working copy cannot be an ancestor of itself. Makes sense?

Reply:
ack, makes sense

Reply:
> Ok, the `descendants()` part breaks `test_next_multiple`

I think that's expected. From memory, that test started something like this:

```
o D
|
o C
|
@ B
|
o A
```

So your working is currently on top of A there. Going one step forward (`jj next`) would put you on top of B, but since that's the current working copy, we skip it and error out. Now, if you were to go two steps forward (`jj next 2`), I think that should work the same way as going forward one step twice (`jj next && jj next`), so I think that should also be an error here. What do you think?

(I know `jj next 2` can be different from `jj next && jj next`, in the case of going across a "diamond" fork+merge.)

Reply:
Ok that seems correct, but still unexpected for simple `next` commands, so both failures are related. It seems the expected behavior is blocked on #229.

Reply:
Why do you think they're blocked on that bug? My feeling is that we shouldn't drop empty commits when auto-rebasing. I think it would rather be something individual commands (like `jj rebase`) would opt in to.

IIUC, after adding that `.descendants()`, the behavior would be as I would expect.

Reply:
As stated, `jj next` moves the working-copy commit forward, but I'd expect everyone to be suprised when `jj next` does not match the behavior of `jj new @-+ & ~@`, which creates a commit on top B. There exists a commit at B, but it shouldn't be the working copy commit after `next`.

Reply:
One alternative is to have `jj next` error out when `@` is not a head. It could print a message like "You're currently editing a non-head. Did you mean `jj next --edit`?".

Reply:
> `jj new @-+ & ~@`, which creates a commit on top B.

Btw, is that actually true? `@-` would be A, so `@-+` would be B, and when subtracting B away, we get an empty set, which would make `jj new` error out. Right?

Reply:
No, just tried it on a local repo. It will create a sibling to the current working copy.

Reply:
Basically `jj new @-+ & ~@` works until the working-copy is not a head commit.

Reply:
Ok, did this time an actual repro with a repo which is similar to the test. `jj new '@-+ & @~'` does indeed not work. I think there are two solutions to solve this problem. Either disallow `next` on non-heads as https://github.com/martinvonz/jj/pull/1200#discussion_r1298652935 suggests or drop the constraint and create a new working copy on top if none exists yet.

[repro.tar.gz](https://github.com/martinvonz/jj/files/12399449/repro.tar.gz)

Reply:
I vote for making it an error

Reply:
This breaks your `jj new` followed by `jj squash` workflow for `jj next` and almost requires `--edit` to be a default.

Reply:
Sorry, I've forgotten what that workflow is. Can you remind me? And why is it broken?

Reply:
It breaks this [workflow](https://github.com/martinvonz/jj/blob/main/docs/github.md#rewriting-commits) and our discussed model, as `next` is no longer able to move the `@` commit around.

Reply:
> It breaks this [workflow](https://github.com/martinvonz/jj/blob/main/docs/github.md#rewriting-commits) and our discussed model in my understanding, as `next` is no longer able to move the `@` commit around.

I still don't follow. There's no `jj next` in that workflow. But if you add one at the end, it would create a new working-copy commit on top of `your-feature`. Isn't that what we want?

Reply:
We briefly discussed this issue in Discord starting [here] and also discovered that currently `next` and `prev` have inconsistent behavior. 

[here]: https://discord.com/channels/968932220549103686/969291218347524238/1143658872775659520

Reply:
I still dislike forcing people into the `edit`ing workflow, there's also a large disclaimer for `jj next --edit` now. I'd prefer `next` to be allowed to `jj co @-+ ~@`, which `prev` does. I think at this point someone with these specific aliases should try out this pr and report back.

Reply:
> I still dislike forcing people into the `edit`ing workflow

I agree. I find it risky to edit "finished" commits. I always create a new commit on top instead.

> I'd prefer `next` to be allowed to `jj co @-+ ~@`

I agree about this too (I have that alias myself and I use it all the time). Isn't that what the code - as implemented - does?

Reply:
> I agree about this too (I have that alias myself and I use it all the time). Isn't that what the code - as implemented - does?

The descendants restriction, leaves an empty set behind, so there nothing to `checkout`. Can you try your alias with `jj co @-+ & ~(@::)` wich matches the current behavior.

Reply:
I didn't create an alias for it but just running `jj co '@-+ & ~(@::)'` on the command line did what I expected. Does it not work for you?

Reply:
Does it still work after an `edit`? If you have a working-copy lying around, the behavior is correct.

Reply:
No, it doesn't work after an `edit`. That doesn't bother me personally because I rarely edit non-head commits. If I did, I think I would want an error if I ran `jj next` without thinking. It would probably also be good on `jj prev`.

Reply:
> If I did, I think I would want an error if I ran `jj next` without thinking.

Why though? IMO it makes sense for Mercurial and Sapling as there is always a commit at the working-copy, but since `jj` generally does not have this requirement, why enforce it. I'd rather allow the creation of a new commit.

Reply:
> So this leaves `--edit` for the known workflow from Mercurial and keeps `next` a Jujutsu like workflow.

I think you meant the other way around. Mercurial doesn't have the working-copy commit feature, so there's no way to edit a non-head commit there (i.e. no way to have commits on top of the working copy). The non-edit workflow is much more like hg and git, where the working copy never has changed on top.

Reply:
> I think you meant the other way around. Mercurial doesn't have the working-copy commit feature, so there's no way to edit a non-head commit there (i.e. no way to have commits on top of the working copy).

This is correct. 

Rephrasing my sentence above, `--edit` is for iterating named commits (the default for Mercurial), while `next` is able to move the empty wc-commit wherever it pleases.

Reply:
Moved the feature request into #2209 to resolve the conversation and land `next/prev`.